### DISCUSSION TITLE
any easy way to add back the / to /auth redirect (need to run it at /auth)

### DISCUSSION THREAD

**Discussion by joeb123456** [keycloak/keycloak#10274] [Archive]
any easy way to add back the / to /auth redirect (need to run it at /auth) do to lots of other config having the full url with /auth in it.

**Comment by Wolfsrudel** [comment]
You can change this back to `/auth` using `http-relative-path`. See the [docs](https://www.keycloak.org/server/all-config?q=http-relative-path&f=build).
So the configuration using the `keycloak.conf` would be
```
http-relative-path=/auth
```
Using `kc.sh`
```
--http-relative-path=/auth
```
And using environment variable
```
KC_HTTP_RELATIVE_PATH="/auth"
```

**Reply by pedroigor** [reply]
Thanks, @Wolfsrudel.

**Reply by Achim-Hentschel** [reply]
Hi @Wolfsrudel

I tried executing `kc.bat` using the following command line (located in the Keycloak home directory on a Windows Server 2019 host).
```
.\bin\kc.bat --verbose start --log-level=INFO --http-relative-path=/auth
``` 

This gives me the following output:
> Unknown options: '--http-relative-path', '/auth'
> Possible solutions: --https-certificate-file, --https-cipher-suites, --https-port, --http-port, --https-key-store-type, --http-enabled, > --https-client-auth, --http-host, --https-protocols, --https-trust-store-file, --https-trust-store-password, --https-trust-store-type, > --https-certificate-key-file, --https-key-store-file, --https-key-store-password
> Try 'kc.bat start --help' for more information on the available options.

Adding it in front of the start does also not solve the issue.

**Reply by pedroigor** [reply]
@Achim-Hentschel The relative path is a build options, available only from the `build` command.

You might want to use the `--auto-build` option as follows to force running a `build` when starting:

```
.\bin\kc.bat start --auto-build --log-level=INFO --http-relative-path=/auth -http-enabled true --hostname-strict false --hostname-strict-https false
``` 

Or just run `start-dev` for dev/testing:

```
kc.bat start-dev --http-relative-path /auth
```

**Reply by Achim-Hentschel** [reply]
Thanks a lot @pedroigor 

Really a pity how blind I sometimes am. I was wondering what the icon in the [All Configurations](https://www.keycloak.org/server/configuration) page means. When I read more carefully through the [Configuring Keycloak](https://www.keycloak.org/server/all-config) documentation again I found this section and finally understood a lot more!
> By looking at the individual configuration options at [All configuration](https://www.keycloak.org/server/all-config) , you notice that some options are marked with an icon to indicate they are a build option. In other words, they can only be set and only take effect when running the build command. You can also check all the options that require a build by looking at the help message of the build command:

I got a suggestion though - I would have understood it on my own if the icon was consistently used/had been introduced on the all config page. I will check the source of the documentation and try contributing that.

**Reply by joeb123456** [reply]
I need /auth but in the old version if landed at / it auto redirected.
now / just shows Resource not found and does not get auto redirected to /auth

**Reply by rssaini01** [reply]
but what about for container image which will be deployed on docker or Kubernetes?

**Reply by sheecegardezi** [reply]
@pedroigor its: --http-enabled true

**Comment by dlmorais-13** [comment]
Hi everyone.

I added the env `KC_HTTP_RELATIVE_PATH` to my Dockerfile prior to build Keycloak and indeed the server is now responding with the `/auth/` suffix. But it's not redirecting.

I have running now in production Keycloak 12.0.4 and the root request responds with a HTML containing the code below:
```html
<script type="text/javascript">
    window.location.href = "/auth/"
</script>
```

I'm trying to migrate to Keycloak 17.0.1 (Quarkus version) and using the request to `/auth/` works as expected, but the root request responds with a `404`.

In my case, as I use a reverse proxy (HAProxy) in front of Keycloak, I can configure this redirect rule in it, but if it's possible to keep the same behavior as Wildfly version it would be very nice.

Is this a case for opening a bug or something?

**Reply by pedroigor** [reply]
Do you mean a request to the welcome page at `http(s)://domain` ?

**Reply by dlmorais-13** [reply]
I mean that requests to `https://domain/` used to "redirect" to `https://domain/auth/`. Here I put "redirect" in quotes because it's a Javascript redirect and not a HTTP redirect.

And now with Quarkus version, even with `KC_HTTP_RELATIVE_PATH` set to `auth`, this "redirect" is not issued.

**Reply by pedroigor** [reply]
Thanks, I see now.

In Wildfly we have an `index.html` as well as a `robots.txt` file. These were served by Wildfly itself.

Do you think it is a must? Otherwise, I would configure the proxy accordingly.

**Reply by dlmorais-13** [reply]
It is a must for my use case, but Keycloak as a product I would only argue in favor of this redirect for the sake of keeping the same behavior as before, making it easier to migrate.

**Reply by joeb123456** [reply]
and changeling the base from /auth to / is to much work / can brake stuff when you have that config over lot's of other systems. vs just one config change in keycloak to go back to the old way.

**Reply by Hildebrand-Ritense** [reply]
Yes, same here. It would really help to have (the option for) a redirect from / to /auth in the Keycloak server itself to keep the existing behaviour and avoid having to set this up separately.

**Reply by AndreiMihaiRad** [reply]
Hi, do you find a solution for automatic redirection on the container level? I'm planning to migrate to KC 18, and it's more difficult to do that on the proxy level at this moment.

**Reply by dlmorais-13** [reply]
@AndreiMihaiRad, depending on your orchestrator, you can add a second container that responds on the root path, issuing the redirect. On k8s you could add a second container to the pod and configure a second pair of service and ingress to handle the root path separately from `/auth` path.

**Reply by rssaini01** [reply]
Don't we have another solution to redirect on `/auth` instead of creating another container to handle redirection?

I don't think it's a good approach. It's having a same issue with 20.0.5

**Reply by shkmaaz11** [reply]
Were you able to find a solution?

**Comment by Sobakaa** [comment]
We have a similar issue with keycloak-x 17.0.1 where some of keycloak's resources are published at /resources/* urls and some, like keycloak.js at auth/js/*.  Our instance is running behind an envoy proxy with the following options:

`kc.sh --verbose build --db postgres --http-relative-path=auth --features=token-exchange`
`kc.sh start --http-enabled=true --hostname-strict=false --hostname-strict-https=false --proxy=passthrough`

Routes are configured as they were in 16.0.1 and everything but admin ui works fine. When trying to access the admin interface, though, user is being redirected to `https://<microservice-name>/auth/` and the fact that we do have a route from "/auth" to "/auth" doesn't help for this one.
Similarly, scripts published by keycloak at /auth/resources/ can be accessed just fine, but keycloak.js at `/auth/js/keycloak.js` would not be returned.
Theres definitely something fishy with the `http-relative-path` that makes it not a 100% replacement for the old way or "redirect".

**Comment by pedroigor** [comment]
@Sobakaa This one in particular should be fixed byhttps://github.com/keycloak/keycloak/issues/11108.

**Comment by beardy247** [comment]
Please stop breaking stuff that's used in production, it makes every upgrade a dice roll. We need to upgrade regularly to keep on top of security fixes, but the threat of breaking changes mean having to do significant regression testing on multiple mobile and web apps every time.

**Reply by pedroigor** [reply]
@beardy247 It is not how it should work, thanks for your feedback. Could you please give us more details about the issues you had during upgrades? Are they related to the distribution and deployment or to the server's core capabilities?

**Reply by beardy247** [reply]
Other than this issue, these are some of the problems we had moving from 16 (wilfdlfy) to 19 (Quarkus): 

- A breaking change was made to the logout redirect url 
- When moving to the Quarkus distribution all the environment variable configuration for the docker container changed radically
- MS SQL Server connection (no signed ssl cert) was broken, required additional connection string param values
- OpenID client scope no longer included by default (https://github.com/keycloak/keycloak/pull/14237)

All of the above required Stack Overflow digging in order to get a resolution and either wasn't mentioned in the upgrade notes or wasn't documented sufficiently to make it straight forward (container config changes)

**Reply by sschu** [reply]
At least the changes with the logout redirect url were mentioned: https://www.keycloak.org/docs/latest/upgrading/index.html#openid-connect-logout-prompt

**Reply by beardy247** [reply]
@sschu It's mentioned, but it's left as an exercise for the reader how to do that when running the containerised version. We could not find any way of doing it with the wildfly version of 18 without having to create our own custom container which we didn't want to do as it defeats the object of having an off the shelf container image to use. This forced us to the Quarkus version which then threw up the other issues.

Keeping up to date with the latest version of Keycloak is important to keep on top of security patches etc so it's vital that upgrading is as painless as possible. The potential for breaking changes means having to regression test every application for every release, which given the number of things we have and the current rate of Keycloak releases is a real challenge. As far as I'm aware only the head release is patched so you've no choice but to move up to stay secure.

**Comment by rmarops** [comment]
Upgraded to keycloak 21 and I am getting "Resource not found" when navigating to keycloak root URL after using `--http-relative-path /auth` (/auth is referenced in all clients)
is the current workaround to use a proxy/load balancer redirect?

**Comment by shkmaaz11** [comment]
> Don't we have another solution to redirect on `/auth` instead of creating another container to handle redirection?
> 
> I don't think it's a good approach. It's having a same issue with 20.0.5
Were you able to find a solution?

**Reply by rmarops** [reply]
unless they offer a built-in redirect in the app, the solution will be external to it.  in my case, AWS load balancer, is the best approach

**Comment by timnolte** [comment]
For anyone running into this issue when running via Docker, and the environment variable doesn't work anymore. If you happen to use Nginx Proxy Manager in front of your Keycloak instance I found that this fix worked for me to get a redirect working again. https://github.com/NginxProxyManager/nginx-proxy-manager/issues/104#issuecomment-1442420244

**Comment by justpalma** [comment]
For anyone using kong as reverse proxy with kubernetes, i did this workaround using a kong puglin
```
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: x-forwarded-header-functions
  namespace: default
config:
  functions:
    - return function() if ngx.var.uri == "/" then kong.service.request.set_path("/auth") end end
plugin: post-function
```

**Comment by deadlysyn** [comment]
nothing works for me, neither environment variable nor build arg when testing 24.0.1. tried each individually and always get "Resource not found" failure to redirect.

```Dockerfile
ENV KC_HTTP_RELATIVE_PATH=/auth
RUN /opt/keycloak/bin/kc.sh build --http-relative-path=auth
RUN /opt/keycloak/bin/kc.sh build --http-relative-path=/auth
RUN /opt/keycloak/bin/kc.sh build --http-relative-path="/auth"
```

**Reply by shkmaaz11** [reply]
@deadlysyn Try 
ENV KC_HOSTNAME_URL=**your_base_url/auth**
or 
RUN /opt/keycloak/bin/kc.sh build --hostname-url=**your_base_url/auth**

I think your hostname url is your_base_url and the relative path of **"/"** is because of that. I've solved my problem by adding the relative path to the baseUrl itself. The http relative auth doesn't work for me either

**Comment by Dahu-Taquin** [comment]
This also causes issues for HSTS : as there is no redirection from "/" to "/auth", no HSTS header is set. Thus, score on some tools like SSLlabs is impacted.

**Comment by mabartos** [comment]
Created an RFE issue: https://github.com/keycloak/keycloak/issues/32863

PR: https://github.com/keycloak/keycloak/pull/32868

Resolved
